connectors/kafka/src/main/java/forklift/controller/KafkaController.java

```diff
@@ -0,0 +1,349 @@
+    }
+
+    private ConsumerRecords<?, ?> flowControlledPoll() throws InterruptedException {
+        //pause partitions that haven't fully been processed yet and unpause those that have
+        Set<TopicPartition> paused = new HashSet<>();
+        Set<TopicPartition> unpaused = new HashSet<>();
+        for (Map.Entry<TopicPartition, AtomicInteger> entry : flowControl.entrySet()) {
+            if (entry.getValue().get() > 0) {
+                paused.add(entry.getKey());
+            } else {
+                unpaused.add(entry.getKey());
+            }
+        }
+        kafkaConsumer.pause(paused);
+        kafkaConsumer.resume(unpaused);
+        if (unpaused.size() == 0 && paused.size() > 0) {
+            synchronized (this) {
+                //wait for flowControl to notify us, resume after a short pause to allow for heartbeats
+                this.wait(100);
+            }
+            return null;
+        } else {
+            return kafkaConsumer.poll(100);
+        }
+    }
+
+    private void updateAssignment() {
+        this.acknowlegmentHandler.addPartitions(kafkaConsumer.assignment());
+        for (TopicPartition partition : kafkaConsumer.assignment()) {
+            this.flowControl.merge(partition, new AtomicInteger(),
+                                   (oldValue, newValue) -> oldValue == null ? newValue : oldValue);
+        }
+    }
+
+    private void addRecordsToStream(ConsumerRecords<?, ?> records) {
+        if (records != null && records.count() > 0) {
+            log.debug("adding: " + records.count() + " to record stream");
+            records.forEach(record -> flowControl.get(new TopicPartition(record.topic(), record.partition()))
+                                                 .incrementAndGet());
+            messageStream.addRecords(consumerRecordsToKafkaMessages(records));
+        }
+    }
+
+    private void commitOffsets(Map<TopicPartition, OffsetAndMetadata> offsets) {
+        if (offsets.size() > 0) {
+            //if the commit fails, perhaps due to a wakeup, store the offsets so they can be retried before the controller stops
+            failedOffset = offsets;
+            kafkaConsumer.commitSync(offsets);
+            failedOffset = null;
+        }
+    }
+
+    private Map<TopicPartition, List<KafkaMessage>> consumerRecordsToKafkaMessages(ConsumerRecords<?, ?> records) {
+        Map<TopicPartition, List<KafkaMessage>> messages = new HashMap<>();
+        records.partitions().forEach(tp -> messages.put(tp, new ArrayList<>()));
+        for (ConsumerRecord<?, ?> record : records) {
+            TopicPartition partition = new TopicPartition(record.topic(), record.partition());
+            messages.get(partition).add(new KafkaMessage(this, record));
+        }
+        return messages;
+    }
+
+    private class RebalanceListener implements ConsumerRebalanceListener {
+        @Override public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
```

Comment:
A little weird to inline the `@Override` here and not elsewhere in the same class (or really anywhere else).